components/gitpod-db/src/typeorm/personal-access-token-db-impl.ts

```diff
@@ -22,7 +22,15 @@ export class PersonalAccessTokenDBImpl implements PersonalAccessTokenDB {
         return (await this.getEntityManager()).getRepository<DBPersonalAccessToken>(DBPersonalAccessToken);
     }
 
-    public async getByHash(hash: string): Promise<DBPersonalAccessToken> {
-        throw new Error("unimplemented");
+    public async getByHash(hash: string): Promise<DBPersonalAccessToken | undefined> {
+        const repo = await this.getRepo();
+        const pat = await repo
+            .createQueryBuilder("pat")
+            .where(`pat.hash = :hash`, { hash })
+            .andWhere(`pat.expirationTime > NOW()`)
```

Comment:
Using `NOW()` for comparison is usually a sign that the code is not very testable, because if you want to test it, you have to rely on side-effects (time passing) that are usually outside of the test writers control.

In other instances we used to pass a now reference, turn that into a string and compare it here.

Reply:
It comes to a trade-off. Ideally, I want to perform that comparison as close to the data source as possible. Especially for something like the PATs. I did consider specifying it pushing it down the the DB as much as possible felt more appropriate.

Even if we supply it, the `getByHash` function actually shouldn't accept it as an argument, we don't really want to specify it arbitrarily. How do we solve it then? Is there something like test clocks in our existing (non-existent) testing setup?

Reply:
Added it optional.